File: maven-root/mtsa/src/main/java/RobotProtocol/src/main/java/robotProtocol/Message.cpp
/*
 * Message.cpp
 *
 *  Created on: Jan 30, 2013
 *      Author: mariano
 */

#include "Message.h"

namespace RobotProtocol {

const char *Message::MESSAGE_HEAD_VALUE		= "LAFHIS";

Message::Message(int messageId, int revision, int to, int from, int flags) {
	this->messageID	= messageId;
	this->revision	= revision;
	this->flags		= flags;
	this->to		= to;
	this->from		= from;
}

Message::Message(char *serializedMessage){
	initialize(serializedMessage);
}

Message::~Message() {
}

int Message::getMessageLength(){
	return MESSAGE_HEADER_LENGTH;
}

int Message::getMessageID(){
	return this->messageID;
}

int Message::getRevision(){
	return this->revision;
}

int Message::getFlags(){
	return this->flags;
}

int Message::getToAddress(){
	return this->to;
}

int Message::getFromAddress(){
	return this->from;
}

void Message::setToAddress(int to){
	this-> to	= to;
}

void Message::setFromAddress(int from){
	this->from	= from;
}

void Message::serializeMessage(char *serializedMessage){
	char **serializedMessageIndex = &serializedMessage;
	writeHeader(serializedMessageIndex);
	writeData(serializedMessageIndex);
}

void Message::initialize(char *serializedMessage){
	char *initialIndex = serializedMessage;
	readHeader(&serializedMessage);
	readData(&serializedMessage);
#ifndef AVR
	if(serializedMessage - initialIndex != getMessageLength())
		throw MESSAGE_LENGTH_ERROR;
#endif
}

void Message::readHeader(char **serializedMessageIndex){
	char *serializedMessage	= *serializedMessageIndex;
	char midField[MID_FIELD_LENGTH + 1];
	char revField[REVISION_FIELD_LENGTH + 1];
	char flagsField[FLAGS_FIELD_LENGTH + 1];
	char toField[TO_FIELD_LENGTH + 1];
	char fromField[FROM_FIELD_LENGTH + 1];
	midField[MID_FIELD_LENGTH] = revField[REVISION_FIELD_LENGTH] = flagsField[FLAGS_FIELD_LENGTH] = '\0';
	toField[TO_FIELD_LENGTH] = fromField[FROM_FIELD_LENGTH] = '\0';
	int index = 0;
	//int messageLength = (int)strtol(*serializedMessageIndex, &next, LENGTH_FIELD_LENGTH);
	index += HEAD_FIELD_LENGTH + LENGTH_FIELD_LENGTH;
	memcpy(midField, &(serializedMessage[index]), MID_FIELD_LENGTH);
	messageID = (int)strtol(midField, NULL, 10);
	index += MID_FIELD_LENGTH;
	memcpy(revField, &(serializedMessage[index]), REVISION_FIELD_LENGTH);
	revision = (int)strtol(revField, NULL, 10);
	index += REVISION_FIELD_LENGTH;
	memcpy(flagsField, &(serializedMessage[index]), FLAGS_FIELD_LENGTH);
	flags = (int)strtol(flagsField, NULL, 10);
	index += FLAGS_FIELD_LENGTH;
	memcpy(toField, &(serializedMessage[index]), TO_FIELD_LENGTH);
	to = (int)strtol(toField, NULL, 10);
	index += TO_FIELD_LENGTH;
	memcpy(fromField, &(serializedMessage[index]), FROM_FIELD_LENGTH);
	from = (int)strtol(fromField, NULL, 10);
	index += FROM_FIELD_LENGTH;
	(*serializedMessageIndex) += index;
	/*
	char* next;
	char midField[MID_FIELD_LENGTH];
	char revField[REVISION_FIELD_LENGTH];
	char flagsField[FLAGS_FIELD_LENGTH];
	int index = 0;

	//int messageLength = (int)strtol(*serializedMessageIndex, NULL, 10);
	(*serializedMessageIndex) += HEAD_FIELD_LENGTH + LENGTH_FIELD_LENGTH;
	memcpy(midField, *serializedMessageIndex, MID_FIELD_LENGTH);
	messageID = (int)strtol(midField, NULL, 10);
	(*serializedMessageIndex) += MID_FIELD_LENGTH;
	memcpy(revField, *serializedMessageIndex, REVISION_FIELD_LENGTH);
	revision = (int)strtol(revField, NULL, 10);
	(*serializedMessageIndex) += REVISION_FIELD_LENGTH;
	memcpy(flagsField, *serializedMessageIndex, FLAGS_FIELD_LENGTH);
	flags = (int)strtol(flagsField, &next, 10);
	(*serializedMessageIndex) += FLAGS_FIELD_LENGTH;
	//int spare = (int)strtol(*serializedMessageIndex, &next, 10);
	(*serializedMessageIndex) += SPARE_FIELD_LENGTH;
	*/
}

void Message::writeHeader(char **serializedMessageIndex){
	sprintf(*serializedMessageIndex,"%s%04d%04d%03d%01d%04d%04d", MESSAGE_HEAD_VALUE, getMessageLength(), messageID, revision, flags, to, from);
	(*serializedMessageIndex) = (*serializedMessageIndex) + MESSAGE_HEADER_LENGTH;
}

} /* namespace RobotProtocol */

File: maven-root/mtsa/src/main/java/MTSAEnactment/ar/uba/dc/lafhis/enactment/robot/N6/robotProtocol/Message.java
package MTSAEnactment.ar.uba.dc.lafhis.enactment.robot.N6.robotProtocol;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

import MTSAEnactment.ar.uba.dc.lafhis.enactment.robot.N6.utils.ReflectionUtils;



public abstract class Message extends MessageDispatcher{

	public static String MESSAGE_HEAD_VALUE = "LAFHIS";

	public static int MESSAGE_LENGTH_ERROR		= -2;
	public static int MESSAGE_END_CHAR_NOT_FOUND= -3;
	public static int MESSAGE_ID_UNKNOWN		= -4;
	public static int MESSAGE_REJECTED			= -5;
	public static int MASTER_UNDEFINED			= -6;	

	public static int MESSAGE_HEADER_LENGTH		= 26;

	public static int HEAD_FIELD_LENGTH			= 6;
	public static int LENGTH_FIELD_LENGTH		= 4;
	public static int MID_FIELD_LENGTH			= 4;
	public static int REVISION_FIELD_LENGTH		= 3;
	public static int FLAGS_FIELD_LENGTH		= 1;
	public static int TO_FIELD_LENGTH			= 4;
	public static int FROM_FIELD_LENGTH			= 4;

	private int messageID;
	private int revision;
	private int flags;
	private int to;
	private int from;	

	public static boolean isMessageComplete(String rawMessage){
		if(rawMessage.length() < MESSAGE_HEADER_LENGTH)
			return false;
		int expectedLength	= getLengthFromRawMessage(rawMessage);
		if(expectedLength > rawMessage.length())
			return false;
		return true;
	}
	
	public static int getLengthFromRawMessage(String rawMessage){
		if(rawMessage.length() < MESSAGE_HEADER_LENGTH)
			return -1;
		return Integer.parseInt(rawMessage.substring(HEAD_FIELD_LENGTH, HEAD_FIELD_LENGTH + LENGTH_FIELD_LENGTH));
	}
	
	public static Message deserializeMessage(String rawMessage){
		int startIndex	= HEAD_FIELD_LENGTH + LENGTH_FIELD_LENGTH;
		int endIndex	= startIndex + MID_FIELD_LENGTH;
		int mid	= Integer.parseInt(rawMessage.substring(startIndex, endIndex));
		List<Class>	subClasses = ReflectionUtils.findSubClasses(Message.class.getPackage().getName(), Message.class.getName()
				, 1, 4);
		subClasses.addAll(ReflectionUtils.findSubClasses(Message.class.getPackage().getName(), Message.class.getName()
				, 1, 4, 911));
		subClasses.addAll(ReflectionUtils.findSubClasses(Message.class.getPackage().getName(), Message.class.getName()
				, 1, 4, 911, 911));
		subClasses.addAll(ReflectionUtils.findSubClasses(Message.class.getPackage().getName(), Message.class.getName()
				, 1, 4, 911, 911, 911));
		subClasses.addAll(ReflectionUtils.findSubClasses(Message.class.getPackage().getName(), Message.class.getName()
				, 1, 4, 911, 911, 911, 911));		
		for(int i = 0; i < subClasses.size(); i++){
			Class c	= (Class)subClasses.get(i);
			try {
				int classMID	= c.getField("MESSAGE_ID").getInt(null);
				if(mid == classMID){
					Constructor<Message> ctor	= c.getDeclaredConstructor(String.class);
					return ctor.newInstance(rawMessage);
				}
			} catch (NoSuchFieldException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (SecurityException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IllegalArgumentException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (IllegalAccessException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (NoSuchMethodException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (InstantiationException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (InvocationTargetException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
		return null;
	} 
	
	protected Message(){
		this(0,0,0,0);
	}
	
	public Message(int messageId, int revision, int to, int from){
		this(messageId, revision, to, from, 0);
	}
	
	public Message(int messageId, int revision, int to, int from, int flags){
		super();
		this.messageID	= messageId;
		this.revision	= revision;
		this.to			= to;
		this.from		= from;
		this.flags		= flags;
	}
	
	public Message(String serializedMessage){
		initialize(serializedMessage);
	}
	
	public final String getPackedMessage(){
		String packedMessage = "";
		packedMessage	+= writeHeader();
		packedMessage	+= writeData();
		return packedMessage;
	}
	
	public int getMessageLength(){
		return MESSAGE_HEADER_LENGTH;
	}

	public int getMessageID(){
		return messageID;
	}

	public int getRevision(){
		return revision;
	}

	public int getFlags(){
		return flags;
	}

	public int getToAddress(){
		return to;
	}

	public int getFromAddress(){
		return from;
	}	

	private void initialize(String inputString){
		int startIndex		= MESSAGE_HEAD_VALUE.length() + LENGTH_FIELD_LENGTH;
		int endIndex		= startIndex + MID_FIELD_LENGTH;
		messageID			= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		startIndex			= endIndex;
		endIndex			= startIndex + REVISION_FIELD_LENGTH;
		revision			= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		startIndex			= endIndex;
		endIndex			= startIndex + FLAGS_FIELD_LENGTH;
		flags				= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		startIndex			= endIndex;
		endIndex			= startIndex + TO_FIELD_LENGTH;
		to					= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		startIndex			= endIndex;
		endIndex			= startIndex + FROM_FIELD_LENGTH;
		from				= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		readData(inputString.substring(endIndex));
	}

	private String writeHeader(){
		return String.format("%s%04d%04d%03d%01d%04d%04d", MESSAGE_HEAD_VALUE, getMessageLength(), messageID, revision, flags, to, from);
	}

	protected abstract void readData(String inputString);
	
	protected abstract String writeData();
	
	protected abstract boolean hasReply();
	
	public void processReply(Message reply){
		fireEvent(new MessageEvent(this, reply));		
	}
}
